x/feeshare/keeper/msg_server.go

```diff
@@ -0,0 +1,226 @@
+package keeper
+
+import (
+	"context"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+
+	"github.com/CosmosContracts/juno/v12/x/feeshare/types"
+)
+
+var _ types.MsgServer = &Keeper{}
+
+// CheckIfDeployerIsContractAdmin ensures the deployer is the contract's admin for all msg_server feeshare functions.
+// If not, it returns an error & we know the user is not authorized to perform the action.
+func (k Keeper) CheckIfDeployerIsContractAdmin(ctx sdk.Context, contract sdk.AccAddress, deployer string) (deployerAddr sdk.AccAddress, errr error) {
+	deployerAddr, err := sdk.AccAddressFromBech32(deployer)
+	if err != nil {
+		return nil, sdkerrors.Wrapf(sdkerrors.ErrInvalidAddress, "invalid deployer address %s", deployer)
+	}
+
+	info := k.wasmKeeper.GetContractInfo(ctx, contract)
+
+	if len(info.Admin) == 0 {
+		return nil, sdkerrors.Wrapf(sdkerrors.ErrUnauthorized, "contract %s has no admin set", contract)
+	}
```

Comment:
Ok, have one discussion point here which I would be curious to get folks thoughts on.

IMO, this should be contract `admin` _or_ `creator` (only if `admin` is not set). Sometimes contracts don't need an `admin`, and as `admin` on contracts has the potential to be abused through contract migrations I don't think we should be incentivizing contracts with admin exclusively.

However, I do think `admin` should take precedence, so would only allow for `creator` to call something like `MsgRegisterFeeshare` if an `admin` is not set.

Thoughts @Reecepbcups? Others?

Reply:
Comment is probably more relevant for line 58 below: https://github.com/CosmosContracts/juno/pull/385/files#diff-81a0948bc3c50667477e54062864d6fa9d845d0d428ea90ffc0386f6ef7a539fR58

Reply:
@JakeHartnell I originally had the logic for creators to have control if there was no admin set. My is that you can't change the `creator`, but you can change the `admin`. But this actually doesn't apply since the whole goal is to not change the person in control. So that is my bad. 

I think it makes sense to add this rather than force admin contracts for fees, which is bad as you alluded too

Reply:
https://github.com/CosmosContracts/juno/pull/385/commits/70d48bbb50397e1ee2218267a7b3d55d3d4ca7ea